handlers/secrets.go

```diff
@@ -75,3 +82,223 @@ func makeSecretsArray(c *client.Client, secretNames []string) ([]*swarm.SecretRe
+	secrets, secretListErr := c.SecretList(context.Background(), types.SecretListOptions{})
+	if secretListErr != nil {
+		return nil, secretListErr
+	}
+
+	for _, secret := range secrets {
+		if secret.Spec.Name == name {
+			return &secret, nil
+		}
+	}
+
+	return nil, fmt.Errorf("not found secret with name: %s", name)
+}
+
+func secretDeleteHandler(c *client.Client, w http.ResponseWriter, r *http.Request) {
+	body, readBodyErr := ioutil.ReadAll(r.Body)
+	if readBodyErr != nil {
+		log.Printf("error deleting secret in secretDeleteHandler: %s\n", readBodyErr)
+
+		w.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	var secret requests.Secret
+
+	unmarshalErr := json.Unmarshal(body, &secret)
+	if unmarshalErr != nil {
+		log.Printf("error unmarshaling secret in secretDeleteHandler: %s\n", unmarshalErr)
+
+		w.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	foundSecret, getSecretErr := getSecretWithName(c, secret.Name)
+	if getSecretErr != nil {
+		log.Printf(
+			"not found secret with name: %s, which you want to remove. Error: %s",
+			secret.Name,
+			getSecretErr,
+		)
+
+		w.WriteHeader(http.StatusNotFound)
+
+		return
+	}
+
+	removeSecretErr := c.SecretRemove(context.Background(), foundSecret.ID)
+	if removeSecretErr != nil {
+		log.Printf(
+			"error trying to remove secret (name: `%s`, ID: `%s`): %s",
+			secret.Name,
+			foundSecret.ID,
+			removeSecretErr,
+		)
+
+		w.WriteHeader(http.StatusInternalServerError)
+
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func secretPutHandler(c *client.Client, w http.ResponseWriter, r *http.Request) {
+	body, readBodyErr := ioutil.ReadAll(r.Body)
+	if readBodyErr != nil {
+		log.Printf("error deleting secret in secretPutHandler: %s\n", readBodyErr)
+
+		w.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	var secret requests.Secret
+
+	unmarshalErr := json.Unmarshal(body, &secret)
+	if unmarshalErr != nil {
+		log.Printf("error unmarshaling secret in secretPutHandler: %s\n", unmarshalErr)
+
+		w.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	foundSecret, getSecretErr := getSecretWithName(c, secret.Name)
+	if getSecretErr != nil {
+		log.Printf(
+			"not found secret with name: %s, which you want to remove. Error: %s",
+			secret.Name,
+			getSecretErr,
+		)
+
+		w.WriteHeader(http.StatusNotFound)
+
+		return
+	}
+
+	nextVersion := swarm.Version{
+		Index: foundSecret.Version.Index + 1,
+	}
+
+	updateSecretErr := c.SecretUpdate(context.Background(), foundSecret.ID, nextVersion, swarm.SecretSpec{
+		Annotations: swarm.Annotations{
+			Name: secret.Name,
+		},
+		Data: []byte(secret.Value),
+	})
+	if updateSecretErr != nil {
+		log.Printf(
+			"couldn't update secret (name: %s, ID: %s) because of error: %s",
+			secret.Name,
+			foundSecret.ID,
+			getSecretErr,
+		)
+
+		w.WriteHeader(http.StatusInternalServerError)
+
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func MakeSecretsHandler(c *client.Client) http.HandlerFunc {
```

Comment:
Looks significantly easier to read. 👍

Reply:
Add on this line:

```
if r.Body !=nil {

    defer r.Body.Close()
}
```

This is needed to close the connection and we have it in every other handler.

Reply:
Added